google/cloud/storage/internal/openssl_util.h

```diff
@@ -35,199 +35,67 @@ namespace storage {
-    std::string b64str = Base64Encode(str);
-    std::replace(b64str.begin(), b64str.end(), '+', '-');
-    std::replace(b64str.begin(), b64str.end(), '/', '_');
-    RightTrim(b64str, '=');
-    return b64str;
-  }
+/**
+ * Signs a string with the private key from a PEM container.
+ *
+ * @return Returns the signature as an *unencoded* byte array. The caller
+ *   might want to use `Base64Encode()` or `HexEncode()` to convert this byte
+ *   array to a format more suitable for transmission over HTTP.
+ */
+std::vector<std::uint8_t> SignStringWithPem(
+    std::string const& str, std::string const& pem_contents,
+    storage::oauth2::JwtSigningAlgorithms alg);
 
- private:
-  static std::unique_ptr<BIO, decltype(&BIO_free_all)>
-  MakeBioChainForBase64Transcoding() {
-    auto base64_io = std::unique_ptr<BIO, decltype(&BIO_free)>(
-        BIO_new(BIO_f_base64()), &BIO_free);
-    auto mem_io = std::unique_ptr<BIO, decltype(&BIO_free)>(
-        BIO_new(BIO_s_mem()), &BIO_free);
-    if (!(base64_io && mem_io)) {
-      std::ostringstream err_builder;
-      err_builder << "Permanent error in " << __func__ << ": "
-                  << "Could not allocate BIO* for Base64 encoding.";
-      google::cloud::internal::ThrowRuntimeError(err_builder.str());
-    }
-    auto bio_chain = std::unique_ptr<BIO, decltype(&BIO_free_all)>(
-        // Output from a b64 encoder should go to an in-memory sink.
-        BIO_push(static_cast<BIO*>(base64_io.release()),
-                 static_cast<BIO*>(mem_io.release())),
-        // Make sure we free all resources in this chain upon destruction.
-        &BIO_free_all);
-    // Don't use newlines as a signal for when to flush buffers.
-    BIO_set_flags(static_cast<BIO*>(bio_chain.get()), BIO_FLAGS_BASE64_NO_NL);
-    return bio_chain;
-  }
+/**
+ * Transforms a string in-place, removing trailing occurrences of a character.
+ *
+ * Warning: this was written with the intent of operating on a string
+ * containing ASCII-encoded (8-bit) characters (e.g. removing trailing '='
+ * characters from a Base64-encoded string) and may not function correctly
+ * for strings containing Unicode characters.
+ */
+inline void RightTrim(std::string& str, char trim_ch) {
+  auto end_pos = str.find_last_not_of(trim_ch);
+  if (std::string::npos != end_pos) str.resize(end_pos + 1);
+}
 
-// The name of the function to free an EVP_MD_CTX changed in OpenSSL 1.1.0.
-#if (OPENSSL_VERSION_NUMBER < 0x10100000L)  // Older than version 1.1.0.
-  inline static std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_destroy)>
-  GetDigestCtx() {
-    return std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_destroy)>(
-        EVP_MD_CTX_create(), &EVP_MD_CTX_destroy);
-  };
-#else
-  inline static std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)>
-  GetDigestCtx() {
-    return std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)>(
-        EVP_MD_CTX_new(), &EVP_MD_CTX_free);
-  };
-#endif
-};
+/**
```

Comment:
How does doxygen deal with only having one description for two functions?

Reply:
It does not, it needs a separate comment, thought this is now moot?